### PR TITLE
Feature pf notification drawer

### PR DESCRIPTION
**Pull request by akoserwal**

### CODE REVIEW

**Comment by akoserwal** [comment]
#146

File: src/app/notification/notification-drawer/example/notification-drawer-example.component.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            </ng-template>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+  <div class="row padding-top-10">
+    <div class="col-sm-12">
+      <div class="panel-group">
+
```
Comment (AdamJ) [inline src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: can you remove this empty line?

File: src/app/notification/notification-drawer/notification-drawer.component.html [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+              <button class="btn btn-link" (click)="onMarkAllRead(notificationGroup)">Mark All Read</button>
+            </span>
+            <span class="drawer-pf-action-link">
+              <button class="btn btn-link" *ngIf="showClearAll" (click)="onClearAll(notificationGroup)">
+                <span class="pficon pficon-close"></span>
+                Clear All
+              </button>
+            </span>
+          </div>
+
```
Comment (AdamJ) [inline src/app/notification/notification-drawer/notification-drawer.component.html]: empty line here

**Comment by akoserwal** [comment]
@mindreeper2420 : done

**Comment by AdamJ** [comment]
@dlabrecq can you review as well? CSS/HTML is good.

**Review by AdamJ — DISMISSED** [review]
CSS/HTML approved.

**Comment by dlabrecq** [comment]
Yes, I plan to review as well. Sorry, been busy with work for summit.

**Comment by dlabrecq** [comment]
Sorry for the delay, been busy with summit work, but I'm starting to review now. I haven't dove into the code yet, but the example looks very nice. Here are a couple quick notes so I don't lose my thought. I'll add more comments as I continue to review.

1. Can you please add an actions area to the example, so we can confirm the notification actions are working as expected? A few other examples have this; toolbar, actions, etc.
2. Can you please add a checkbox to show what empty state looks like?
3. For next time, please setup a live example to run via rawgit? Adam and I are able to run the example locally, so not a big deal, but not all designers have that ability. Screenshots are also very helpful.

Instructions below.
https://github.com/patternfly/patternfly-ng/blob/master/CONTRIBUTING.md

File: src/app/notification/notification-drawer/notification-drawer.component.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    </ng-template>
+    <ng-template #showTitleTemplate [ngTemplateOutlet]="titleTemplate"></ng-template>
+  </div>
+
+
+  <div *ngIf="!notificationGroups" class="apf-blank-notification-groups">
+    <pfng-empty-state [config]="emptyStateConfig"></pfng-empty-state>
+  </div>
+
+  <div *ngIf="notificationGroups" pf-fixed-accordion scroll-selector=".panel-body">
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/notification-drawer.component.html]: How are pf-fixed-accordion and scroll-selector being used here? I don't see any new directives in this PR. In fact, the drawer still appears to render properly if I remove them

File: src/app/notification/notification-drawer/notification-drawer.component.html [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    <a class="drawer-pf-close pficon pficon-close" (click)="onClose()"></a>
+    <div *ngIf="titleTemplate; then showTitleTemplate else showTitle"></div>
+    <ng-template #showTitle>
+      <h3 class="text-center">{{title}}</h3>
+    </ng-template>
+    <ng-template #showTitleTemplate [ngTemplateOutlet]="titleTemplate"></ng-template>
+  </div>
+
+
+  <div *ngIf="!notificationGroups" class="apf-blank-notification-groups">
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/notification-drawer.component.html]: apf-blank-notification-groups seems to be a non-existing selector?

File: src/app/notification/notification-drawer/example/notification-drawer-example.component.html [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        </tab>
+        <tab heading="html">
+          <include-content src="src/app/notification/notification-drawer/example/notification-drawer-example.component.html"></include-content>
+        </tab>
+        <tab heading="typescript">
+          <include-content src="src/app/notification/notification-drawer/example/notification-drawer-example.component.ts"></include-content>
+        </tab>
+      </tabset>
+    </div>
+  </div>
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: Seems to be missing an ending div tag

File: src/app/notification/notification-drawer/example/notification-drawer-example.component.html [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+              <h3 class="text-center">Notifications Drawer</h3>
+            </ng-template>
+            <ng-template #headingTemplate let-group="group" let-index="index">
+              {{group.heading}}
+            </ng-template>
+            <ng-template #subHeadingTemplate let-group="group" let-index="index">
+              <span>{{group.subHeading}}</span>
+            </ng-template>
+            <ng-template #notificationBodyTemplate let-notify="notify" let-index="index">
+              <div uib-dropdown class="dropdown pull-right dropdown-kebab-pf" *ngIf="notify.moreActions.length > 0">
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: It seems only the pull-right selector us needed here? I was able to remove uib-dropdown, dropdown, and dropdown-kebab-pf.

Note that we're not using the old uib-dropdown directive for dropdowns. If you need a dropdown, please use ngx-boostrap.

File: src/app/notification/notification-drawer/example/notification-drawer-example.component.html [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                <pfng-action class="list-pf-actions" [config]="actionConfig"></pfng-action>
+              </div>
+              <span class="pull-left pficon pficon-ok" *ngIf="notify.type === 'success'"></span>
+              <span class="pull-left pficon pficon-info" *ngIf="notify.type === 'info'"></span>
+              <span class="pull-left pficon pficon-error-circle-o" *ngIf="notify.type === 'danger'"></span>
+              <span class="pull-left pficon pficon-warning-triangle-o" *ngIf="notify.type === 'warning'"></span>
+              <div class="drawer-pf-notification-content">
+                <span class="drawer-pf-notification-message" tooltip-append-to-body="true" tooltip-popup-delay="500" tooltip-placement="bottom"
+                  tooltip="notify.message">
+                  {{notify.message}}
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: Do any of the examples show tooltips for you? I see tooltips for the kebab, but not here.

File: src/app/notification/notification-drawer/notification-drawer.component.html [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            <span class="drawer-pf-action-link">
+              <button class="btn btn-link" *ngIf="showClearAll" (click)="onClearAll(notificationGroup)">
+                <span class="pficon pficon-close"></span>
+                Clear All
+              </button>
+            </span>
+          </div>
+          <div *ngIf="!hasNotifications(notificationGroup)">
+            <div class="panel-body">
+              <pfng-empty-state [config]="emptyStateConfig"></pfng-empty-state>
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/notification-drawer.component.html]: I see that the empty state config is used in two places. I'm wondering if we should have a unique empty state for each notification and one just for the notification drawer itself? If one isn't provided, perhaps we just apply the one for the notification drawer. Thoughts?

File: src/app/notification/notification-drawer/notification-drawer.component.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+  hasUnread(group: NotificaitonGroup) {
+    return size(filter(get(group, 'notifications'), { 'isViewing': false })) > 0;
+  }
+
+
+  /**
+   *  Return boolean if group has notifications
+   *  @param group 
+   */
+  hasNotifications(group: NotificaitonGroup[]) {
```
Comment (dlabrecq) [inline src/app/notification/notification-drawer/notification-drawer.component.ts]: Please add return types to the methods. For example:

hasNotifications(group: NotificaitonGroup[]): boolean
onClose(): void
...

File: src/app/notification/notification-group.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+   *  
+   */
+  heading: string;
+  
+  
+  /**
+   * sub-heading area for each notification group
+   * 
+   */
+  subHeading: string;
```
Comment (dlabrecq) [inline src/app/notification/notification-group.ts]: If this isn't required, please add a '?' char

**Review by dlabrecq — CHANGES_REQUESTED** [review]
Looks great! Easy to review. Just minor comments

File: src/app/notification/notification-drawer/notification-drawer.component.html (hunk 3 above)
Comment (akoserwal) [reply src/app/notification/notification-drawer/notification-drawer.component.html]: @dlabrecq : I'll remove them.

File: src/app/notification/notification-drawer/notification-drawer.component.html (hunk 4 above)
Comment (akoserwal) [reply src/app/notification/notification-drawer/notification-drawer.component.html]: @dlabrecq : Sure, I'll remove it

File: src/app/notification/notification-drawer/example/notification-drawer-example.component.html (hunk 5 above)
Comment (akoserwal) [reply src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: @dlabrecq : oh, yes!

Comment (akoserwal) [reply src/app/notification/notification-drawer/example/notification-drawer-example.component.html]: done

File: src/app/notification/notification-drawer/notification-drawer.component.html (hunk 8 above)
Comment (akoserwal) [reply src/app/notification/notification-drawer/notification-drawer.component.html]: @dlabrecq: For unique empty state, we can have empty state config property defined in NotificationGroup and one emptyconfig for the notification drawer.

Comment (dlabrecq) [reply src/app/notification/notification-drawer/notification-drawer.component.html]: That sounds good

**Comment by akoserwal** [comment]
@dlabrecq : please can you review now.

**Comment by jeff-phillips-18** [comment]
The notification drawer should scroll the inner body of the notification groups rather than the entire drawer. That is, similar to the fixed height accordion, all the groups should be visible at all times and the scrolling should occur within the open group.

This is done a couple of different ways. In PF core and in Angular Patternfly, there is javascript used to compute and set the height of the body, listening for window resize events as well to update this height. In the patternfly-react repo flexbox is used to set the heights appropriately.

I see that there is a use of `pf-fixed-accordion` at https://github.com/patternfly/patternfly-ng/pull/334/files#diff-c7e7858ac8154ee94820361e24979b53R17 which appears to be desiring to use the fixed accordion which has not yet been implemented here.

**Comment by akoserwal** [comment]
@jeff-phillips-18: thanks, I'll work on the scroll part.

**Comment by dlabrecq** [comment]
Thank you for making the requested changes -- looking good. Could you please update the actions text area when the kebab options are selected?

**Comment by akoserwal** [comment]
@jeff-phillips-18 : scroll should be visible only for notification tabs which have loading spinner ?

**Comment by jeff-phillips-18** [comment]
The scroll should be in the group body whenever there are more notifications than can be included without pushing the following group headers out of view. This functionality can be seen at https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?selectedKind=NotificationDrawer&selectedStory=Notification%20Drawer&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

**Comment by akoserwal** [comment]
@jeff-phillips-18 : got it, thanks

**Comment by akoserwal** [comment]
@jeff-phillips-18 : please can you check the flexbox approach picked from patternfly-react.

**Comment by akoserwal** [comment]
@jeff-phillips-18: thank you for your help. @dlabrecq: please can you review now.

**Review by dlabrecq — APPROVED** [review]
Built and tested locally. Confirmed that the notification drawer now scrolls the inner body of each notification group.

**Review by jeff-phillips-18 — APPROVED** [review]

**Pull request merged**